### PR TITLE
ticketbuyer: Add RPC calls to get/set config

### PR DESCRIPTION
**Pull request by tuxcanfly**
As mentioned in #463 we need a way to tweak ticketbuyer config  over the RPC. This PR implements RPCs to allow get/set on commonly used config fields.

### CODE REVIEW

File: rpc/rpcserver/server.go [hunk 1]
```diff
@@ -134,6 +134,181 @@ func decodeAddress(a string, params *chaincfg.Params) (dcrutil.Address, error) {
...
+		AvgPriceMode:          config.AvgPriceMode,
+		AvgPriceVWAPDelta:     int64(config.AvgPriceVWAPDelta),
+		BalanceToMaintain:     int64(config.BalanceToMaintainAbsolute),
+		BlocksToAvg:           int64(config.BlocksToAvg),
+		DontWaitForTickets:    config.DontWaitForTickets,
+		ExpiryDelta:           int64(config.ExpiryDelta),
+		FeeSource:             config.FeeSource,
+		FeeTargetScaling:      config.FeeTargetScaling,
+		MinFee:                int64(config.MinFee),
+		MinPriceScale:         config.MinPriceScale,
```
Comment (alexlyp) [inline rpc/rpcserver/server.go]: This is gone. Should be removed

File: rpc/rpcserver/server.go [hunk 2]
```diff
@@ -134,6 +134,181 @@ func decodeAddress(a string, params *chaincfg.Params) (dcrutil.Address, error) {
...
+		MaxFee:                int64(config.MaxFee),
+		MaxPerBlock:           int64(config.MaxPerBlock),
+		MaxPriceAbsolute:      int64(config.MaxPriceAbsolute),
+		MaxPriceRelative:      config.MaxPriceRelative,
+		MaxPriceScale:         config.MaxPriceScale,
```
Comment (alexlyp) [inline rpc/rpcserver/server.go]: This is deprecated, and should be removed

File: rpc/api.proto [hunk 3]
```diff
@@ -481,3 +491,75 @@ message StartAutoBuyerResponse {}
...
+    string avg_price_mode = 2;
+    int64 avg_priceVWAP_delta = 3;
+    int64 balance_to_maintain = 4;
+    int64 blocks_to_avg = 5;
+    bool dont_wait_for_tickets = 6;
+    int64 expiry_delta = 7;
+    string fee_source = 8;
+    double fee_target_scaling = 9;
+    int64 min_fee = 10;
+    double min_price_scale = 11;
```
Comment (alexlyp) [inline rpc/api.proto]: remove

File: rpc/api.proto [hunk 4]
```diff
@@ -481,3 +491,75 @@ message StartAutoBuyerResponse {}
...
+    int64 max_fee = 12;
+    int64 max_per_block = 13;
+    int64 max_price_absolute = 14;
+    double max_price_relative = 15;
+    double max_price_scale = 16;
```
Comment (alexlyp) [inline rpc/api.proto]: remove

File: ticketbuyer/buyer.go [hunk 5]
```diff
@@ -67,8 +68,8 @@ type Config struct {
 	ExpiryDelta               int
 	FeeSource                 string
 	FeeTargetScaling          float64
-	HighPricePenalty          float64
 	MinFee                    float64
+	MinPriceScale             float64
```
Comment (alexlyp) [inline ticketbuyer/buyer.go]: Remove

File: ticketbuyer/buyer.go [hunk 6]
```diff
@@ -107,6 +105,248 @@ type TicketPurchaser struct {
...
+		AvgPriceMode:              t.cfg.AvgPriceMode,
+		AvgPriceVWAPDelta:         t.cfg.AvgPriceVWAPDelta,
+		BalanceToMaintainAbsolute: t.balanceToMaintain,
+		BlocksToAvg:               t.cfg.BlocksToAvg,
+		DontWaitForTickets:        t.cfg.DontWaitForTickets,
+		ExpiryDelta:               t.cfg.ExpiryDelta,
+		FeeSource:                 t.cfg.FeeSource,
+		FeeTargetScaling:          t.cfg.FeeTargetScaling,
+		MinFee:                    t.cfg.MinFee,
+		MinPriceScale:             t.cfg.MinPriceScale,
```
Comment (alexlyp) [inline ticketbuyer/buyer.go]: remove

File: ticketbuyer/buyer.go [hunk 7]
```diff
@@ -484,7 +732,7 @@ func (t *TicketPurchaser) Purchase(height int64) (*PurchaseStats, error) {
 		// Estimated number of tickets to be bought with stake reward
 		tixToBuyWithStakeRewardFunds := stakeRewardFunds / nextStakeDiff.ToCoin()
 		// Amount of tickets that can be bought with existing funds
-		tixCanBuy := (bal.Spendable.ToCoin() - balanceToMaintainAmt.ToCoin()) / nextStakeDiff.ToCoin()
+		tixCanBuy := (bal.Spendable.ToCoin() - balanceToMaintainAmt) / nextStakeDiff.ToCoin()
```
Comment (alexlyp) [inline ticketbuyer/buyer.go]: Why this change?

File: ticketbuyer/buyer.go [hunk 8]
```diff
@@ -448,19 +696,19 @@ func (t *TicketPurchaser) Purchase(height int64) (*PurchaseStats, error) {
...
 		log.Infof("Not enough funds to buy tickets: (spendable: %v, balancetomaintain: %v) ",
-			bal.Spendable.ToCoin(), balanceToMaintainAmt.ToCoin())
+			bal.Spendable.ToCoin(), balanceToMaintainAmt)
 	}
 
 	// For spreading your ticket purchases evenly throughout window.
 	// Use available funds to calculate how many tickets to buy, and also
 	// approximate the income you're going to have from older tickets that
-	// you've voted and are maturing during this window
+	// you've voted and are maturing during this window (tixWillRedeem)
```
Comment (alexlyp) [inline ticketbuyer/buyer.go]: Why this change?

File: ticketbuyer/buyer.go [hunk 9]
```diff
@@ -448,19 +696,19 @@ func (t *TicketPurchaser) Purchase(height int64) (*PurchaseStats, error) {
 	log.Tracef("Ticket allotment left in window is %v, blocks left is %v",
 		ticketsLeftInWindow, (int(winSize) - t.idxDiffPeriod))
 
-	toBuyForBlock := int(math.Floor((bal.Spendable.ToCoin() - balanceToMaintainAmt.ToCoin()) / nextStakeDiff.ToCoin()))
+	toBuyForBlock := int(math.Floor((bal.Spendable.ToCoin() - balanceToMaintainAmt) / nextStakeDiff.ToCoin()))
```
Comment (alexlyp) [inline ticketbuyer/buyer.go]: Why this change?

File: ticketbuyer/buyer.go [hunk 10]
```diff
@@ -576,7 +824,7 @@ func (t *TicketPurchaser) Purchase(height int64) (*PurchaseStats, error) {
 	// have enough moneys.
 	notEnough := func(bal dcrutil.Amount, toBuy int, sd dcrutil.Amount) bool {
 		return (bal.ToCoin() - float64(toBuy)*sd.ToCoin()) <
-			balanceToMaintainAmt.ToCoin()
+			balanceToMaintainAmt
```
Comment (alexlyp) [inline ticketbuyer/buyer.go]: another ToCoin() change

Comment (tuxcanfly) [inline ticketbuyer/buyer.go]: Ok, the `float` -> `dcrutil.Amount` changes have been reverted.

File: rpc/rpcserver/server.go (hunk 1 above)
Comment (tuxcanfly) [inline rpc/rpcserver/server.go]: Removed.

File: rpc/api.proto (hunk 3 above)
Comment (tuxcanfly) [inline rpc/api.proto]: Removed.

File: ticketbuyer/buyer.go (hunk 9 above)
Comment (tuxcanfly) [inline ticketbuyer/buyer.go]: Reverted.

File: rpc/rpcserver/server.go [hunk 11]
```diff
@@ -134,6 +134,232 @@ func decodeAddress(a string, params *chaincfg.Params) (dcrutil.Address, error) {
...
+	poolAddress := req.PoolAddress
+	poolFees := pm.Purchaser().PoolFees()
+
+	switch {
+	case poolFees == 0 && poolAddress != "":
+		return nil, grpc.Errorf(codes.InvalidArgument, "Pool address set but no pool fees given")
+	case poolFees != 0 && poolAddress == "":
+		return nil, grpc.Errorf(codes.InvalidArgument, "Pool fees set but no pool address given")
+	case poolFees != 0 && poolAddress != "":
+		err = txrules.IsValidPoolFeeRate(poolFees)
```
Comment (tuxcanfly) [inline rpc/rpcserver/server.go]: Guess this check is not really required here.

Comment (tuxcanfly) [inline rpc/rpcserver/server.go]: Fixed.
